### PR TITLE
Improve DAU estimation quality in stats and p3a

### PR DESCRIPTION
**Pull request by keur**
## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by keur** [comment]
sigh rebased wrong :/

File: browser/BUILD.gn
Comment (bridiver) [inline browser/BUILD.gn:455]: can we move this stuff into browser/brave_stats/BUILD.gn (and move the files as well)? This BUILD.gn is already big and cumbersome so it would be good to start moving things out when we touch them

File: browser/brave_stats_tab_helper.cc
Comment (bridiver) [inline browser/brave_stats_tab_helper.cc:10]: we're missing deps for content/public/browser in `stats_updater` target

File: browser/brave_stats_tab_helper.cc
Comment (bridiver) [inline browser/brave_stats_tab_helper.cc:29]: you're missing a header for the transition types (and also the corresponding dep for `//ui/base`

Comment (bridiver) [inline browser/brave_stats_tab_helper.cc:29]: lint nit: `case` should indented

File: browser/brave_stats_tab_helper.h
Comment (bridiver) [inline browser/brave_stats_tab_helper.h:11]: this should be in the .cc file

File: browser/brave_stats_updater.cc
Comment (bridiver) [inline browser/brave_stats_updater.cc:10]: there's a missing dep for `//base` in `stats_updater`

File: browser/brave_stats_updater.cc
Comment (bridiver) [inline browser/brave_stats_updater.cc:19]: pref_names.h has `assert(enable_brave_referrals)` so this header and any pref names it references need to be behind a buildflag

File: common/network_constants.cc
Comment (bridiver) [inline common/network_constants.cc:11]: this should be defined in config.js/declare_args

File: browser/brave_stats_updater.cc
Comment (bridiver) [inline browser/brave_stats_updater.cc:79]: use command line flags to override instead of env and pass the full url. I think you only need one flag to override the url and not two

File: browser/brave_stats_updater.cc
Comment (bridiver) [inline browser/brave_stats_updater.cc:15]: also missing dep for brave/common, brave/components/brave_referrals/buildflags and brave/components/brave_referrals

File: browser/brave_stats_updater.cc
Comment (bridiver) [reply browser/brave_stats_updater.cc:19]: assuming this stuff is even needed anymore since you removed this https://github.com/brave/brave-core/pull/6660/files#diff-1f3055c17e07e8133098b79725256804L84

Comment (bridiver) [reply browser/brave_stats_updater.cc:19]: looks like the prefs are already behind a buildflag https://github.com/brave/brave-core/pull/6660/files#diff-1f3055c17e07e8133098b79725256804R192

Comment (bridiver) [reply browser/brave_stats_updater.cc:19]: also make sure you add `//brave/components/brave_referrals/common` inside and `enable_brave_referrals` guard in gn

File: browser/brave_stats_updater.cc
Comment (bridiver) [inline browser/brave_stats_updater.cc:249]: if you're building this dynamically I would add a DCHECK to make sure it's a valid gurl

File: browser/brave_stats_updater.cc
Comment (bridiver) [inline browser/brave_stats_updater.cc:171]: why do you have `(void)` here?

File: browser/brave_stats_updater.h
Comment (bridiver) [inline browser/brave_stats_updater.h:54]: nit: I would return a GURL here

File: browser/brave_stats_updater.cc
Comment (bridiver) [inline browser/brave_stats_updater.cc:123]: using base/one_shot_event might simplify this stuff

File: browser/brave_stats_tab_helper.cc
Comment (emerick) [inline browser/brave_stats_tab_helper.cc:34]: I would add a `break` here just to play it safe.

File: browser/brave_stats_updater_browsertest.cc
Comment (bridiver) [inline browser/brave_stats_updater_browsertest.cc:75]: if this env is only for tests then I would add a SetEndpointForTesting method instead of the command line

File: browser/brave_stats_updater.cc
Comment (bridiver) [reply browser/brave_stats_updater.cc:19]: if anything is behind a buildflag in gn, it also needs a buildflag in c++ and vice versa https://github.com/brave/brave-core/pull/6660/files#diff-84f8700538180eec90bb2f4c9dff4da4R474

File: common/network_constants.cc
Comment (bridiver) [reply common/network_constants.cc:11]: just like https://github.com/brave/brave-core/pull/6660/files#diff-84f8700538180eec90bb2f4c9dff4da4R452

File: browser/brave_stats_updater_browsertest.cc
Comment (bridiver) [reply browser/brave_stats_updater_browsertest.cc:75]: also there should be some new tests here to cover the new functionality

File: browser/brave_stats_updater.cc
Comment (keur) [reply browser/brave_stats_updater.cc:171]: it returns a value. the return value is actually used in `brave_stats_tab_helper`. this is just calling that method back in case the score was >= `kMinimumUsageThreshold`, but was not triggered because the first usage ping hadn't fired yet. it's just making it clear that it's calling that back but not using the return value.

Comment (bridiver) [reply browser/brave_stats_updater.cc:177]: are you getting a compiler error? You should normally only have a problem if the method has `WARN_UNUSED_RESULT` in which case you'd use ignore_result from base/macros.h. While using `(void)` here isn't technically wrong, it's just not standard practice in chromium code

File: browser/BUILD.gn
Comment (bridiver) [inline browser/BUILD.gn:174]: nit: if you name the target "brave_stats" then you can just use "brave_stats" without any `:`
not a big deal, but that's the standard convention unless you need to expose multiple targets inside brave_stats

File: browser/brave_local_state_prefs.cc
Comment (bridiver) [inline browser/brave_local_state_prefs.cc:46]: nit: `brave_stats::RegisterLocalStatePrefs`
as far as conventions go, `RegisterPrefsForAdBlockService` is wrong and `ntp_background_images::NTPBackgroundImagesService::RegisterLocalStatePrefs` is correct

File: browser/brave_stats/BUILD.gn
Comment (bridiver) [inline browser/brave_stats/BUILD.gn:5]: We're leaving these empty by default in both gn and config.js now because we don't want other browsers based on Brave to hit these endpoints. Contact @mihaiplesa about adding this to CI

**Comment by iefremov** [comment]
do we have a ticket/spec for this?

File: browser/brave_stats/switches.cc
Comment (iefremov) [inline browser/brave_stats/switches.cc:15]: you can make it `constexpr` and put into the header entirely

**Comment by iefremov** [comment]
sorry, I'll finish the review tomorrow

File: browser/brave_stats/BUILD.gn
Comment (bridiver) [inline browser/brave_stats/BUILD.gn:19]: looks like you're missing some deps for this
`//services/network/public/cpp`, `//net`, maybe others. You should have a dep for every header file you use

File: browser/brave_stats/BUILD.gn
Comment (bridiver) [inline browser/brave_stats/BUILD.gn:41]: you can't put the buildflags in `enable_brave_referrals` because you need access to it when the buildflag is disabled

File: browser/brave_stats/brave_stats_updater.cc
Comment (bridiver) [inline browser/brave_stats/brave_stats_updater.cc:32]: this should be alphabetized above

File: build/commands/lib/config.js
Comment (bridiver) [inline build/commands/lib/config.js:115]: same as gn config, do not include default values for the url

File: components/p3a/brave_p3a_service.cc
Comment (bridiver) [inline components/p3a/brave_p3a_service.cc:23]: at some point we need to fix this layer violation cc @iefremov

File: components/webcompat_reporter/browser/BUILD.gn
Comment (bridiver) [inline components/webcompat_reporter/browser/BUILD.gn:26]: this looks like it was a missing dep before, but this is also a layering violation we need to fix. Can we move these `brave_stats` methods into `//brave/components/brave_stats` with the exception of `GetFirstRunTime` because that would also create a layering violation with chrome/browser/first_run

File: components/p3a/brave_p3a_service.cc
Comment (iefremov) [reply components/p3a/brave_p3a_service.cc:23]: right, I think at some point it will be fixed during our `gn check` campaign

File: common/pref_names.cc
Comment (iefremov) [inline common/pref_names.cc:19]: `chrome/common/pref_names.cc` has a good tradition of documenting prefs - can we put some explanations here as well?

File: browser/brave_stats/brave_stats_updater.h
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.h:53]: `SetStatsThresholdCallbackForTests` ?

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:83]: can you please `clang-format` the whole thing?

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:37]: pls make these constants constexpr and put them into anonymous namespace (enclosed into `brave_stats` namespace)

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:80]: you can initialize it in the header file

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:86]: mb write a LOG(ERROR) if this is an invalid url? handy for manual testing

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:108]: I believe this DCHEK is pretty obvious, so can be dropped

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:236]: ?

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:275]: no need to test the timer

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:310]: can we reduce the boilerplate by moving comon parts into a function?

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:279]: Does it really need to be a class method? I think this is just a free function that should live in the anomymous namespace

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:48]: move this to unnamed namespace as well

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:328]: maybe this is `SendUserTriggeredPing`? Just want to have more understandable name

**Review by iefremov — APPROVED** [review]
Approving to unblock merge - in general looks good, but please fix highlighted nits

File: browser/brave_stats/BUILD.gn
Comment (keur) [reply browser/brave_stats/BUILD.gn:5]: blocked by https://github.com/brave/devops/issues/3731

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:99]: you can merge these two namespaces :)

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:49]: no need to add C-style `static` for functions, anonymous namespace does the trick

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:40]: if you add "seconds" to the variable name, it would be much more descriptive on its own

File: browser/brave_stats/brave_stats_updater.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater.cc:33]: pls add a whitespace before

File: browser/brave_stats/brave_stats_updater_browsertest.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater_browsertest.cc:116]: clang-format

File: browser/brave_stats/brave_stats_updater_params.cc
Comment (iefremov) [inline browser/brave_stats/brave_stats_updater_params.cc:16]: clang-format (sorting)

**Review by bridiver — APPROVED** [review]

**Pull request merged**
